syntax = "proto3";

package proto;

service TweetService {
	// Tweet a message.
	rpc Tweet(TweetIn) returns (TweetOut) {}
	// Follow a user.
	rpc Follow(FollowIn) returns (FollowOut) {}
	// Show tweet from a followy or yourself.
	rpc Show(ShowIn) returns (ShowOut) {}
	// Login to twitter.
	rpc Login(LoginIn) returns (LoginOut) {}
	// Logout the current user.
	rpc Logout(LogoutIn) returns (LogoutOut) {}
	// Register a new user.
	rpc Register(RegisterIn) returns (RegisterOut) {}
}

message TweetIn {
	string user = 1;
	string content = 2;
	int64 time = 3;
}

message TweetOut {
	bool error = 1;
}

message FollowIn {
	string name = 1;
}

message FollowOut {
	bool error = 1;
}

message ShowIn {
	string user = 1;
}

message ShowOut {
	repeated TweetIn tweets = 1;
	bool error = 2;
}

message LoginIn {
	string user = 1;
	string pass = 2;
}

message LoginOut {
	bool error = 1;
}

message LogoutIn {
	// nothing.
}

message LogoutOut {
	bool error = 1;
}

message RegisterIn {
	string name = 1;
	string pass = 2;
}

message RegisterOut {
	bool error = 1;
}
